parser/src/cfg/chord.rs

```diff
@@ -10,105 +15,328 @@ pub(crate) fn parse_defchordv2(
+        if !all_participating_key_sets.insert(chord.participating_keys) {
+            ParseError::new_without_span(
                 "This chord has previously been defined.\n\
-                Only one set of chords must exist for one key combination."
+                Only one set of chords must exist for one key combination.",
             );
+        } else {
+            for pkey in chord.participating_keys.iter().copied() {
+                //log::trace!("chord for key:{pkey:?} > {chord:?}");
+                chords_container
+                    .mapping
+                    .entry(pkey)
+                    .or_insert(ChordsForKey { chords: vec![] })
+                    .chords
+                    .push(s.a.sref(chord.clone()));
+            }
         }
+    }
+    let rem = chunks.remainder();
+    if !rem.is_empty() {
+        bail_expr!(
+            rem.last().unwrap(),
+            "Incomplete chord entry. Each chord entry must have 5 items:\n\
+        particpating-keys, action, timeout, release-type, disabled-layers"
+        );
+    }
+    Ok(chords_container)
+}
 
-        let action = parse_action(&chunk[1], s)?;
-        let timeout = parse_non_zero_u16(&chunk[2], s, "chord timeout")?;
-        let release_behaviour = chunk[3]
-            .atom(s.vars())
-            .and_then(|r| {
-                Some(match r {
-                    "first-release" => ReleaseBehaviour::OnFirstRelease,
-                    "all-released" => ReleaseBehaviour::OnLastRelease,
-                    _ => return None,
+fn parse_single_chord(chunk: &[SExpr], s: &ParserState) -> Result<ChordV2<'static, KanataCustom>> {
+    let keys = &chunk[0];
+    let key_strings = keys
+        .list(s.vars())
+        .map(|keys| {
+            keys.iter()
+                .map(|key| {
+                    key.atom(s.vars()).ok_or_else(|| {
+                        anyhow_expr!(key, "The first chord item must be a list of keys.")
+                    })
                 })
+                .collect::<Vec<_>>()
+        })
+        .ok_or_else(|| anyhow_expr!(keys, "The first chord item must be a list of keys."))?
+        .into_iter()
+        .collect::<Result<Vec<_>>>()?;
+    if key_strings.len() < 2 {
+        bail_expr!(keys, "The minimum number of participating chord keys is 2");
+    }
+    let participants = parse_participating_keys(key_strings)?;
+    let action = parse_action(&chunk[1], s)?;
+    let timeout = parse_timeout(&chunk[2], s)?;
+    let release_behaviour = parse_release_behaviour(&chunk[3], s)?;
+    let disabled_layers = parse_disabled_layers(&chunk[4], s)?;
+    let chord: ChordV2<'static, KanataCustom> = ChordV2 {
+        action,
+        participating_keys: s.a.sref_vec(participants.clone()),
+        pending_duration: timeout,
+        disabled_layers: s.a.sref_vec(disabled_layers),
+        release_behaviour,
+    };
+    Ok(s.a.sref(chord).clone())
+}
+
+fn parse_participating_keys(key_strings: Vec<&str>) -> Result<Vec<u16>> {
```

Comment:
The new code is collecting into a `Vec` 3 times (when including `keys->key_strings` code above) to get participating keys where before it was a single `try_fold`. The optimizer can _probably_ reuse the allocation, and it _is_in the parser so it's not critical to get the most performance - but what's the rationale behind the code change for parsing participating keys?

Reply:
I had first tried a different approach where the chord file parsing did not happen via an artificial `SExpr` and instead called the extracted methods with `String`s.
From a software engineering perspective this imo would be preferable to creating a full S-Expression and then immediately parsing it into a `SExpr`, but it was too much effort and not worth it in the end.
Changed